### PR TITLE
Adding new movies feature

### PR DESCRIPTION
**Pull request by matej494**

### CODE REVIEW

File: Movieator/App/Movie list/MovieListViewController.swift [hunk 1]
```diff
@@ -27,11 +28,36 @@ class MovieListViewController: UIViewController {
...
+        let alert = UIAlertController(title: "Add new movies", message: "", preferredStyle: .alert)
+        let importButton = UIAlertAction(title: "Import", style: .default, handler: { action in
+            if let id = alert.textFields?.first?.text {
+                self.importMovie(with: id)
+            }
+        })
+        alert.addAction(importButton)
+        alert.addTextField { textField in
+            textField.placeholder = "Place IMDB ID here."
+            NotificationCenter.default.addObserver(forName: nil, object: textField, queue: OperationQueue.main, using: { _ in
```
Comment (dkulundzic) [inline Movieator/App/Movie list/MovieListViewController.swift]: I actually never seen this approach used - creative 😄

File: Movieator/App/Movie list/MovieListViewController.swift [hunk 2]
```diff
@@ -139,17 +139,25 @@ private extension MovieListViewController {
...
                 self.present(alert, animated: true) },
             failure: { error in
-                let alert = UIAlertController(title: "Movie not saved", message: "Error ocured while trying to fetch movie", preferredStyle: .alert)
-                alert.addAction(UIAlertAction(title: "Ok", style: .default, handler: nil))
+                let alert = UIAlertController(title: "Movie not found", message: error.errorDescription, preferredStyle: .alert)
+                alert.addAction(UIAlertAction(title: "Ok", style: .cancel, handler: nil))
                 self.present(alert, animated: true) })
     }
+    
+    func importMovie(for movie: Movie) {
```
Comment (dkulundzic) [inline Movieator/App/Movie list/MovieListViewController.swift]: Remove all redundant `self` references within this method.

File: Movieator/App/Movie list/MovieListViewController.swift [hunk 3]
```diff
@@ -139,17 +139,25 @@ private extension MovieListViewController {
...
+        self.data.saveMovies(movie: movie)
+        let alert = UIAlertController(title: "Movie saved", message: "", preferredStyle: .actionSheet)
```
Comment (dkulundzic) [inline Movieator/App/Movie list/MovieListViewController.swift]: An `.actionSheet` style doesn't make much sense here, as it's purely an info alert.

**Comment by dkulundzic** [comment]
@matej494, one bonus point task that we could do here. There's a lot of code repetition surrounding the `UIAlertController` - do you see any approaches that we could take to refactor the code into something reusable?

**Comment by matej494** [comment]
I can't think of any way to solve this problem in a good way because every instance of `UIAlertController` is different.

**Comment by dkulundzic** [comment]
They are different, but they always require the same input parameters for configuration:
- Optional title
- Optional message
- Style
- Actions

Does this help a bit?

**Comment by matej494** [comment]
I've figured that out, but at best case I managed to replace two lines with one, and the original code was easier to read and understand.

**Comment by dkulundzic** [comment]
Alright. Tomorrow we'll check out something that might be useful here.

File: Movieator.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -20,6 +20,7 @@
 		F5337BEA20B309F3004C6E77 /* SavedMoviesManager.swift in Sources */ = {isa = PBXBuildFile; fileRef = F5337BE920B309F3004C6E77 /* SavedMoviesManager.swift */; };
 		F5337BEC20B30F48004C6E77 /* SavedMovie.swift in Sources */ = {isa = PBXBuildFile; fileRef = F5337BEB20B30F48004C6E77 /* SavedMovie.swift */; };
 		F555BDB720AAEE43004466D3 /* MovieListViewController.swift in Sources */ = {isa = PBXBuildFile; fileRef = F555BDB620AAEE43004466D3 /* MovieListViewController.swift */; };
+		F566D9E620BE8EA200B1D197 /* UIAlertControllerExtension-Movieator.swift in Sources */ = {isa = PBXBuildFile; fileRef = F566D9E520BE8EA200B1D197 /* UIAlertControllerExtension-Movieator.swift */; };
```
Comment (dkulundzic) [inline Movieator.xcodeproj/project.pbxproj]: We don't need `Extension` in `UIAlertControllerExtension-Movieator.swift` - just `UIAlertController-Movieator.swift` is great.

File: Movieator/App/Movie list/MovieListViewController.swift [hunk 5]
```diff
@@ -56,26 +56,24 @@ class MovieListViewController: UIViewController {
...
+                        }),
+                       UIAlertAction(title: "Release date", style: .default, handler: { [weak self] action in
+                        self?.sortMovies(withKey: "releaseDate")
+                        }),
+                       UIAlertAction(title: "IMDB rating", style: .default, handler: { [weak self] action in
+                        self?.sortMovies(withKey: "imdbRating")
+                        }),
+                       UIAlertAction(title: "Metascore rating", style: .default, handler: { [weak self] action in
+                        self?.sortMovies(withKey: "metascore")
+                        })]
```
Comment (dkulundzic) [inline Movieator/App/Movie list/MovieListViewController.swift]: How about formatting the actions creation like this:
```
let actions = [
            UIAlertAction(title: "Title", style: .default, handler: { [weak self] action in
                self?.sortMovies(withKey: "title")
            }),
            UIAlertAction(title: "Release date", style: .default, handler: { [weak self] action in
                self?.sortMovies(withKey: "releaseDate")
            }),
            UIAlertAction(title: "IMDB rating", style: .default, handler: { [weak self] action in
                self?.sortMovies(withKey: "imdbRating")
            }),
            UIAlertAction(title: "Metascore rating", style: .default, handler: { [weak self] action in
                self?.sortMovies(withKey: "metascore")
            })
]
```

Looks quite more readable to me.

File: Movieator/App/Movie list/MovieListViewController.swift [hunk 6]
```diff
@@ -144,20 +142,17 @@ private extension MovieListViewController {
...
     }
     
     func importMovie(for movie: Movie) {
-        self.data.saveMovies(movie: movie)
-        let alert = UIAlertController(title: "Movie saved", message: "", preferredStyle: .actionSheet)
-        alert.addAction(UIAlertAction(title: "Ok", style: .cancel, handler: nil))
-        self.present(alert, animated: true)
+        data.saveMovies(movie: movie)
+        let alert = UIAlertController.generic(title: "Movie saved", cancelTitle: "Ok")
+        alert.present(target: self)
```
Comment (dkulundzic) [inline Movieator/App/Movie list/MovieListViewController.swift]: The `UIAlertController` call-site looks pretty good, well done!

File: Movieator/Common/Extensions/UIAlertControllerExtension-Movieator.swift [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+//  UIAlertControllerExtension-Movieator.swift
+//  Movieator
+//
+//  Created by Matej Korman on 30/05/2018.
+//  Copyright © 2018 Codeopolius. All rights reserved.
+//
+
+import UIKit
+
+extension UIAlertController {
```
Comment (dkulundzic) [inline Movieator/Common/Extensions/UIAlertControllerExtension-Movieator.swift]: Good, good 👍

File: Movieator/Common/Extensions/UIAlertControllerExtension-Movieator.swift [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    static func generic(title: String?,
+                        message: String? = "",
+                        preferredStyle: UIAlertControllerStyle = .alert,
+                        actions: [UIAlertAction] = [],
+                        cancelTitle: String = "Cancel") -> UIAlertController {
+        let alertController = UIAlertController(title: title, message: message, preferredStyle: preferredStyle)
+        actions.forEach { action in
+            alertController.addAction(action)
+        }
+        if !actions.contains { $0.style == .cancel } {
```
Comment (dkulundzic) [inline Movieator/Common/Extensions/UIAlertControllerExtension-Movieator.swift]: Shorthand notation, ei? Very nice 🙂

File: Movieator/Common/Extensions/UIAlertControllerExtension-Movieator.swift [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+            alertController.addAction(UIAlertAction(title: cancelTitle, style: .cancel, handler: nil))
+        }
+        return alertController
+    }
+    
+    func present(target: UIViewController, animated: Bool = true) {
```
Comment (dkulundzic) [inline Movieator/Common/Extensions/UIAlertControllerExtension-Movieator.swift]: What do you think about renaming the `target` argument to `on`? It's a bit more readable and clear that way, at least to me.

**Comment by dkulundzic** [comment]
Feel free to proceed with the merge.

**Comment by matej494** [comment]
Successfully merged branch `add-movies` to `develop` without conflicts.

**Pull request merged**